client/src/push-content.ts

```diff
@@ -47,14 +75,46 @@ export const getMessage = async (): Promise<string | undefined> => {
 
 export const pushContent = () => async () => {
   if (await canPush(getErrorDiagnosticsBySource())) {
-    await _pushContent(getRepo, getMessage, vscode.window.showInformationMessage, vscode.window.showErrorMessage)()
+    await _pushContent(getRepo, taggingDialog, getNewTag, getMessage, vscode.window.showInformationMessage, vscode.window.showErrorMessage)()
   }
 }
 
-export const _pushContent = (_getRepo: () => Repository, _getMessage: () => Thenable<string | undefined>, infoReporter: (msg: string) => Thenable<string | undefined>, errorReporter: (msg: string) => Thenable<string | undefined>) => async () => {
+export const _pushContent = (
+  _getRepo: () => Repository,
+  _taggingDialog: () => Thenable<string | undefined>,
+  _getTag: (repo: Repository, release: boolean) => string | undefined,
+  _getMessage: () => Thenable<string | undefined>,
+  infoReporter: (msg: string) => Thenable<string | undefined>,
+  errorReporter: (msg: string) => Thenable<string | undefined>
+) => async () => {
   const repo = _getRepo()
   const commitOptions: CommitOptions = { all: true }
 
+  const tagging = await _taggingDialog()
+  let tag: string | undefined
+  if (tagging !== 'No Tag') {
+    /* instanbul ignore next */
+    const tagMode = tagging === 'Release'
+    tag = _getTag(repo, tagMode)
+  }
+
+  if (tag !== undefined) {
+    /* instanbul ignore next */
+    try {
+      await (repo as any)._repository.tag(tag) // when VSCode API is updated -> await repo.tag(tag)
+    } catch (e) {
+      const message: string = e.gitErrorCode === undefined ? e.message : e.gitErrorCode
+      void errorReporter(`Tagging failed: ${message} ${String(e.stderr)}`)
+    }
+    return
+  }
+
+  const diff = await repo.diffWithHEAD()
```

Comment:
🤔 - Is there a case where we'd want to apply a tag but there's no diff? I'm wondering if this check should actually go into `canPush()` (or at least earlier in this flow) if we always want to do nothing when there's no local changes?

Reply:
It looks like tagging and committing will stay separate, since a user may commit and then later decide to tag.

Reply:
Oh, actually a follow up question regarding workflow: If a user wants to commit + tag on a push, doesn't the tagging need to happen after creating the commit, or am I misunderstanding something here?

Reply:
No misunderstanding.